### ISSUE TITLE
Improvements to secrets support

### ISSUE THREAD

**Issue by samalba** [dagger/dagger#4366]
This serves as an index for all the issues related to secrets. If someone takes ownership on this issue, this should start by triaging the issues list below, corresponding to a proposal to add support for secrets.

### What is supported

As of Dagger Engine 0.4.3, a secret can be...

- [loaded from an Environment variable set on the host machine](https://pkg.go.dev/dagger.io/dagger@v0.4.3#HostVariable.Secret)
- [injected into a container as an environment variable](https://pkg.go.dev/dagger.io/dagger@v0.4.3#Container.WithSecretVariable)
- [injected into a container as a file](https://pkg.go.dev/dagger.io/dagger@v0.4.3#Container.WithMountedSecret)

### What is currently not supported

As of Dagger Engine 0.4.3, the following limitations or problems exist:

```[tasklist]
### Tasks
- [ ] #4413 
- [ ] #4426 
- [ ] #4289
- [ ] #3323
- [ ] https://github.com/dagger/dagger/issues/4403
- [ ] https://github.com/dagger/dagger/issues/4339
- [ ] https://github.com/dagger/dagger/issues/4864
- [ ] https://github.com/dagger/dagger/issues/4896
```

**Comment by shykes** [comment]
> * 0.3 does not support secret creation from a container

I think this is wrong, that is supported with `file { secret }`.

**Comment by aluzzardi** [comment]
Correct — creating a secret from a container is supported, but leaks the secret in the container image and cache.

Also, not supported is creating a secret in code (#3442)

**Comment by samalba** [comment]
My bad, I did not know that was possible already. However, as @aluzzardi pointed out this isn't secure, not sure we should keep the support. I corrected the description above.

When I mount a secret to a container, this is safe as it won't end up in my logs or cache layers. As user, I would expect this to be true for a secret created from a container, without having to deal with the underlying complexity.

**Comment by shykes** [comment]
> My bad, I did not know that was possible already. However, as @aluzzardi pointed out this isn't secure, not sure we should keep the support. I corrected the description above.
> 
> 
> 
> When I mount a secret to a container, this is safe as it won't end up in my logs or cache layers. As user, I would expect this to be true for a secret created from a container, without having to deal with the underlying complexity.

I agree, IMO we should at least discuss how to fix the security issue. Here's my earlier comment about this: https://github.com/dagger/dagger/issues/4339#issuecomment-1376110970

**Comment by sagikazarmark** [comment]
Just to clarify: creating a secret from a container is what's "unsafe" at the moment, right? 

Creating a secret from an env var or even by talking to directly to Vault to fetch the secret value and then passing that in the container as an env var should be "safe" (ie. won't be stored anywhere unencrypted), right?

**Comment by shykes** [comment]
> Just to clarify: creating a secret from a container is what's "unsafe" at the moment, right?
> 
> Creating a secret from an env var or even by talking to directly to Vault to fetch the secret value and then passing that in the container as an env var should be "safe" (ie. won't be stored anywhere unencrypted), right?

@sagikazarmark, the following methods **are considered safe**:

* Load a secret from an environment variable on the host. [Here's an example](https://play.dagger.cloud/playground/zM-OU1mQcWC). NOTE: only variables set *before opening the Dagger session* are available in this way.
* Injecting a secret into a container as an environment variable, using the `withSecretVariable` operation. [Here's an example](https://play.dagger.cloud/playground/5C9LZEBUmks)
* Injecting a secret into a container as a file. [Here's an example](https://play.dagger.cloud/playground/7bDtWKnx3i9).

The big inconvenience at the moment, is that the env var method is not dynamic: you can only access env variables that were set *at the time the dagger session was open*. In your case for example, you must talk to Vault, and export your secrets as environment variables *before opening the dagger session*.

TLDR: there are safe ways to load secrets into Dagger, but they are not dynamic.

**Comment by sagikazarmark** [comment]
Let's see if I got this right: if I want to use a secret from a secret store, I have to fetch that secret **before** I call `dagger.Connect` (in Go) and set it as an environment variable manually (regardless if I want to access it as an env var or a file in the container).

Is that correct?

**Comment by jlongtine** [comment]
> Let's see if I got this right: if I want to use a secret from a secret store, I have to fetch that secret **before** I call `dagger.Connect` (in Go) and set it as an environment variable manually (regardless if I want to access it as an env var or a file in the container).
> 
> Is that correct?

Yes, this is correct. At least for our current setup. The intent is to change that soon, so more dynamic secrets are possible.

**Comment by shykes** [comment]
FYI I created an individual issue for the cache leak problem: https://github.com/dagger/dagger/issues/4403

Fixing that would be the shortest path to safe, dynamic secrets creation. But we don't know (yet) how difficult it would be to implement.

**Comment by marcindulak** [comment]
> Just to clarify: creating a secret from a container is what's "unsafe" at the moment, right?
> 
> Creating a secret from an env var or even by talking to directly to Vault to fetch the secret value and then passing that in the container as an env var should be "safe" (ie. won't be stored anywhere unencrypted), right?

I see the examples from https://github.com/dagger/dagger/issues/4366#issuecomment-1387485982 are in GraphQL, but in practical terms, aren't the secrets from environment variables in a container stored also unencrypted in `/proc/<pid>/environ`?

```sh
docker run --rm -e SECRET=TEST --name bash bash bash -c "cat /proc/1/environ | grep SECRET"
```

**Comment by shykes** [comment]
> > Just to clarify: creating a secret from a container is what's "unsafe" at the moment, right?
> 
> > 
> 
> > Creating a secret from an env var or even by talking to directly to Vault to fetch the secret value and then passing that in the container as an env var should be "safe" (ie. won't be stored anywhere unencrypted), right?
> 
> 
> 
> I see the examples from https://github.com/dagger/dagger/issues/4366#issuecomment-1387485982 are in GraphQL, but in practical terms, aren't the secrets from environment variables in a container stored also unencrypted in `/proc/<pid>/environ`?
> 
> 
> 
> ```sh
> 
> docker run --rm -e SECRET=TEST --name bash bash bash -c "cat /proc/1/environ | grep SECRET"
> 
> ```

Yes. We consider that to be safe, since it's commonly accepted best practice, regardless of Dagger. Your secrets will be stored unencrypted in memory, accessible only to your process & superuser.

If you want to go above and beyond, eg. selinux/apparmor lockdown, secure enclave, dedicated hardware, it's not explicitly supported but we could explore it together. Do you have a specific mechanism in mind?

**Comment by sagikazarmark** [comment]
> > Let's see if I got this right: if I want to use a secret from a secret store, I have to fetch that secret **before** I call `dagger.Connect` (in Go) and set it as an environment variable manually (regardless if I want to access it as an env var or a file in the container).
> > Is that correct?
> 
> Yes, this is correct. At least for our current setup. The intent is to change that soon, so more dynamic secrets are possible.

Thanks for the clarification!

**Comment by marcosnils** [comment]
> * NOTE: only variables set _before opening the Dagger session_ are available in this way.

Would it make sense to add to support to the session binary to update its env variables? Since we already have an IPC pipe between the SDK and the session binary, adding logic in the SDK's to allow dynamic variables doesn't seem that difficult?

**Comment by sipsma** [comment]
> Would it make sense to add to support to the session binary to update its env variables? Since we already have an IPC pipe between the SDK and the session binary, adding logic in the SDK's to allow dynamic variables doesn't seem that difficult? .

The IPC pipe is just http requests over localhost from sdk clients to the dagger session server. If we want the session to be able to request data from the SDKs we'd need to support bidirectional communication. Multiple possibilities (can setup a websocket between the sdk+session, could have the sdk also listen on localhost, could do rpc over pipes, etc.) but they'll all require a decent bit of effort afaik and that effort will be multiplied by the number of SDKs. So IMO it's totally possible but the cost in terms of time and maintenance will be significant. If there are other use cases for this sort of bidirectional communication then it would be more compelling.

---

Personally, I think that just putting an explicit secret secretion API where the secret is just in-memory and tied to the lifetime of the session is probably easiest and lowest hanging fruit: https://github.com/dagger/dagger/issues/3442

**Comment by marcosnils** [comment]
> If we want the session to be able to request data from the SDKs we'd need to support bidirectional communication.

Can't the SDK's just push the env updates to the session binary? Do we need bidirectional communication for that?

**Comment by sipsma** [comment]
We'd need either
1. a mechanism for knowing when env vars are updated and an api in the session for updating them
2. or the ability to intercept host env var graphql calls and then update their values in the session when they are made

I don't know entirely how to implement `1.`. `2` is probably possible but feels like a can of worms since I don't know how to integrate it cleanly with the codegen code, I'd probably prefer bidirectional communication over that.

**Comment by sipsma** [comment]
> I was thinking something along this line. 

I think one extra tricky part will be the need to handle the case where a secret env var is referenced inside of an ID, in which case it won't hit the hooks we plug into the SDK code. That's technically solvable, but it would be ugly and it would rely on being able to deserialize those IDs in the SDKs, but then if we ever change the IDs in the future (the fact that they are base64 encoded LLB is just an implementation detail, not to be relied on) then we need to update them, etc. That's where the can of worms comes from I think.

**Comment by marcosnils** [comment]
> I think one extra tricky part will be the need to handle the case where a secret env var is referenced inside of an ID

:thinking: but in order to get the `SecretID` from the env var you needed to call `withSecretVariable` in the first place? How could a user craft a request with a secret env var without calling the SDK `withSecretVariable` in the query build chain?

**Comment by sipsma** [comment]
> 🤔 but in order to get the SecretID from the env var you needed to call withSecretVariable in the first place? How could a user craft a request with a secret env var without calling the SDK withSecretVariable in the query build chain?

Right, the corner cases that remain are:
1. Technically, users can use IDs between sessions. They really shouldn't in theory but we actually do this in our docs sometimes so it's sort of a "that ship has sailed" situation. We'd have to decide whether we care about this gotcha or not.
2. If the env var changes at any point it won't get re-loaded with the new value. I don't think this is a very common use case at all, but it would be another gotcha lurking.

Those aren't necessarily dealbreakers, but it does seem like we'd have to do a fair bit (setup sdk codegen hooks, setup a separate api in the session binary for updating the env vars) and then at the end we'd just get the ability to update env vars after `dagger.Connect` but with a bunch of caveats. I'm not sure if it's worth it relative to the other options available.

**Comment by marcosnils** [comment]
> Right, the corner cases that remain are:
> 
> 1. Technically, users can use IDs between sessions. They really shouldn't in theory but we actually do this in our docs sometimes so it's sort of a "that ship has sailed" situation. We'd have to decide whether we care about this gotcha or not.
> 2. If the env var changes at any point it won't get re-loaded with the new value. I don't think this is a very common use case at all, but it would be another gotcha lurking.

Agree, I briefly thought about those and kind of ignored them since as you mentioned, they're kind of "exceptional" cases. 


> and then at the end we'd just get the ability to update env vars after `dagger.Connect` but with a bunch of caveats.

Also agree, I just thought it is a cheaper way to support dynamic secrets until we have the secrets store solution. Doing something like this (https://discord.com/channels/707636530424053791/1003718839739105300/1062063056626143242) would be very useful IMO.

**Comment by d3rp3tt3** [comment]
Out of this list, the security issue is the most pressing to address: https://github.com/dagger/dagger/issues/4366#.

**Comment by samalba** [comment]
> Out of this list, the security issue is the most pressing to address: https://github.com/dagger/dagger/issues/4366.

Did you mean #4339?

**Comment by d3rp3tt3** [comment]
Yes! I meant #4339 and #4413.

**Comment by d3rp3tt3** [comment]
@samalba , I updated the original summary to use a task list, so it's easier for us to track.

**Comment by mircubed** [comment]
Thank you to everyone who provided input for these improvements and to @dolanor for all the work! You can read about the improvements here - https://dagger.io/blog/dagger-0-5-0
